File: python/src/shapes/client/shell/runner.py
from collections.abc import Callable
from typing import Any, final

from shapes.client.shell import handler

from .input_reader import read_input


@final
class Runner:
    def __init__(self):
        print("=" * 53)
        print("Welcome to the Area and Perimeter Calculator!")
        print("=" * 53)

    @staticmethod
    def starting_message():
        print("Choose a shape to calculate its area and perimeter.")
        print("  [C] CIRCLE")
        print("  [R] RECTANGLE")
        print("  [S] SQUARE")
        print("Type '[E]XIT' to quit the program.")

    def run(self):
        # pylint: disable=unnecessary-lambda-assignment
        print_thank_you_message = lambda: (
            print("Thank you for using the Shapes CLI!"),
            print("P.S. Did you know? The secret code is 42!"),
        )

        while True:
            Runner.starting_message()

            shape = read_input()
            if shape == "q":
                print_thank_you_message()
                break

            actions: dict[str, Callable[..., Any]] = {
                "circle": handler.handle_circle,
                "c": handler.handle_circle,
                "rectangle": handler.handle_rectangle,
                "r": handler.handle_rectangle,
                "square": handler.handle_square,
                "s": handler.handle_square,
            }

            action = actions.get(
                shape,
                lambda: (
                    print("Invalid input. Please try again."),
                    print("Type one of the listed options."),
                ),
            )

            action()

            print("=" * 53)

File: ruby/lib/shapes/client/shell/runner.rb
# frozen_string_literal: true

require_relative "handler"
require_relative "input_reader"

class Runner
    def initialize
        puts "=" * 53
        puts "Welcome to the Area and Perimeter Calculator!"
        puts "=" * 53
    end

    def self.starting_message
        puts "Choose a shape to calculate its area and perimeter."
        puts "  [C] CIRCLE"
        puts "  [R] RECTANGLE"
        puts "  [S] SQUARE"
        puts "Type '[E]XIT' to quit the program."
    end

    def run
        print_thank_you_message = proc do
            puts "Thank you for using the Shapes CLI!"
            puts "P.S. Did you know? The secret code is 42!"
        end

        loop do
            Runner.starting_message

            print "Enter a shape (or 'q' to quit): "
            shape = gets.strip.downcase

            if shape == "q"
                print_thank_you_message.call
                break
            end

            actions = {
                "circle" => -> { Handler.handle_circle },
                "c" => -> { Handler.handle_circle },
                "rectangle" => -> { Handler.handle_rectangle },
                "r" => -> { Handler.handle_rectangle },
                "square" => -> { Handler.handle_square },
                "s" => -> { Handler.handle_square }
            }

            action = actions.fetch(shape, lambda {
                puts "Invalid input. Please try again."
                puts "Type one of the listed options."
            })

            action.call

            puts "=" * 50
        end
    end
end
